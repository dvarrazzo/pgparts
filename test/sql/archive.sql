\pset format unaligned
\pset footer off
\set VERBOSITY terse
create schema arctest;
set datestyle = 'ISO';

create extension pgparts with schema arctest;

create table at (
    id serial primary key,
    day date not null,
    data text);

-- Create some partitions and data
select arctest.setup('at', 'day', 'monthly');
select arctest.create_for('at', '2017-01-10');
select arctest.create_for('at', '2017-02-10');
select arctest.create_for('at', '2017-03-10');
insert into at (day) values ('2017-01-10');
insert into at (day) values ('2017-02-10');

-- Must create the archive before
select arctest.archive_before('at', '2017-02-15');

select arctest.create_archive('at');
select * from at order by id;
select * from at_all order by id;
select * from at_archived order by id;

-- Idempotent
select arctest.create_archive('at');
select * from at order by id;
select * from at_all order by id;
select * from at_archived order by id;

-- Archiving partitions
select arctest.archive_before('at', '2017-02-15');
select * from arctest.info('at', '2017-01-15');
select * from arctest.info('at', '2017-02-15');
select * from at order by id;
select * from at_all order by id;
select * from at_archived order by id;

-- Archived partition can't receive data
insert into at (day) values ('2017-01-10');
insert into at (day) values ('2017-02-10');
select * from at order by id;
select * from at_all order by id;
select * from at_archived order by id;

-- Can't re-create an archived partition
select arctest.create_for('at', '2017-01-10');

-- Archived partition can be unarchived
select arctest.unarchive_partition('at_201701');
select arctest.unarchive_partition('at_201702');
select * from arctest.info('at', '2017-01-15');
select * from arctest.info('at', '2017-02-15');

-- Unarchived partition can receive data
insert into at (day) values ('2017-01-10');
insert into at (day) values ('2017-02-10');
select * from at order by id;
select * from at_all order by id;
select * from at_archived order by id;

drop extension pgparts;
