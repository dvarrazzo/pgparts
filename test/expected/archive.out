\pset format unaligned
\pset footer off
\set VERBOSITY terse
create schema arctest;
set datestyle = 'ISO';
create extension pgparts with schema arctest;
create table at (
    id serial primary key,
    day date not null,
    data text);
-- Create some partitions and data
select arctest.setup('at', 'day', 'monthly');
setup

select arctest.create_for('at', '2017-01-10');
NOTICE:  creating partition public.at_201701 of table at
create_for
at_201701
select arctest.create_for('at', '2017-02-10');
NOTICE:  creating partition public.at_201702 of table at
create_for
at_201702
select arctest.create_for('at', '2017-03-10');
NOTICE:  creating partition public.at_201703 of table at
create_for
at_201703
insert into at (day) values ('2017-01-10');
insert into at (day) values ('2017-02-10');
-- Must create the archive before
select arctest.archive_before('at', '2017-02-15');
ERROR:  archive table for at not found
select arctest.create_archive('at');
NOTICE:  creating table at_all
NOTICE:  creating table at_archived
create_archive
at_archived
select * from at order by id;
id|day|data
1|2017-01-10|
2|2017-02-10|
select * from at_all order by id;
id|day|data
1|2017-01-10|
2|2017-02-10|
select * from at_archived order by id;
id|day|data
-- Idempotent
select arctest.create_archive('at');
create_archive
at_archived
select * from at order by id;
id|day|data
1|2017-01-10|
2|2017-02-10|
select * from at_all order by id;
id|day|data
1|2017-01-10|
2|2017-02-10|
select * from at_archived order by id;
id|day|data
-- Archiving partitions
select arctest.archive_before('at', '2017-02-15');
NOTICE:  archiving partition at_201701
archive_before
at_201701
select * from arctest.info('at', '2017-01-15');
state|partition
archived|at_201701
select * from arctest.info('at', '2017-02-15');
state|partition
present|at_201702
select * from at order by id;
id|day|data
2|2017-02-10|
select * from at_all order by id;
id|day|data
1|2017-01-10|
2|2017-02-10|
select * from at_archived order by id;
id|day|data
1|2017-01-10|
-- Archived partition can't receive data
insert into at (day) values ('2017-01-10');
ERROR:  partition public.at_201701 missing for day = '2017-01-10'
insert into at (day) values ('2017-02-10');
select * from at order by id;
id|day|data
2|2017-02-10|
4|2017-02-10|
select * from at_all order by id;
id|day|data
1|2017-01-10|
2|2017-02-10|
4|2017-02-10|
select * from at_archived order by id;
id|day|data
1|2017-01-10|
-- Can't re-create an archived partition
select arctest.create_for('at', '2017-01-10');
ERROR:  the partition at_201701 exists but was archived
-- Archived partition can be unarchived
select arctest.unarchive_partition('at_201701');
NOTICE:  unarchiving partition at_201701
unarchive_partition
at_201701
select arctest.unarchive_partition('at_201702');
ERROR:  The table at_201702 is not an archived partition
select * from arctest.info('at', '2017-01-15');
state|partition
present|at_201701
select * from arctest.info('at', '2017-02-15');
state|partition
present|at_201702
-- Unarchived partition can receive data
insert into at (day) values ('2017-01-10');
insert into at (day) values ('2017-02-10');
select * from at order by id;
id|day|data
1|2017-01-10|
2|2017-02-10|
4|2017-02-10|
5|2017-01-10|
6|2017-02-10|
select * from at_all order by id;
id|day|data
1|2017-01-10|
2|2017-02-10|
4|2017-02-10|
5|2017-01-10|
6|2017-02-10|
select * from at_archived order by id;
id|day|data
drop extension pgparts;
